src/schedule_systems.rs

```diff
@@ -85,75 +82,105 @@ pub(crate) fn run_synctest<C: Config>(world: &mut World, mut sess: SyncTestSessi
+
+    match requests {
+        Some(Ok(requests)) => handle_requests(requests, world),
+        Some(Err(GGRSError::PredictionThreshold)) => {
+            info!("P2PSpectatorSession: Waiting for input from host.")
+        }
+        Some(Err(e)) => warn!("{e}"),
+        None => {}
+    };
 }
 
 pub(crate) fn run_p2p<C: Config>(world: &mut World, mut sess: P2PSession<C>) {
-    maybe_init_snapshots(world, sess.max_prediction());
-
     world.insert_resource(LocalPlayers(sess.local_player_handles()));
 
-    if sess.current_state() == SessionState::Running {
+    let running = sess.current_state() == SessionState::Running;
+
+    if running {
         // get local player inputs
         world.run_schedule(ReadInputs);
+
         let local_inputs = world.remove_resource::<LocalInputs<C>>().expect(
             "No local player inputs found. Did you insert systems into the ReadInputs schedule?",
         );
+
         for (handle, input) in local_inputs.0 {
             sess.add_local_input(handle, input)
                 .expect("All handles in local_inputs should be valid");
         }
-
-        match sess.advance_frame() {
-            Ok(requests) => handle_requests(requests, world),
-            Err(GGRSError::PredictionThreshold) => {
-                info!("Skipping a frame: PredictionThreshold.")
-            }
-            Err(e) => warn!("{e}"),
-        };
     }
 
+    let requests = running.then(|| sess.advance_frame());
+
     world.insert_resource(Session::P2P(sess));
+
+    match requests {
+        Some(Ok(requests)) => handle_requests(requests, world),
+        Some(Err(GGRSError::PredictionThreshold)) => {
+            info!("Skipping a frame: PredictionThreshold.")
+        }
+        Some(Err(e)) => warn!("{e}"),
+        None => {}
+    }
 }
 
 pub(crate) fn handle_requests<T: Config>(requests: Vec<GGRSRequest<T>>, world: &mut World) {
     for request in requests {
+        let current_frame = world
+            .get_resource::<RollbackFrameCount>()
+            .map(|frame| frame.0)
+            .unwrap_or_default();
+
+        let confirmed_frame = match world.get_resource::<Session<T>>() {
+            Some(Session::P2P(s)) => Some(s.confirmed_frame()),
+            Some(Session::SyncTest(s)) => {
+                // TODO: `max_prediction` should be replaced with `depth`, but it is private.
+                let current_frame = current_frame - (s.max_prediction() as i32);
+                (current_frame < 0).then_some(current_frame)
+            }
+            Some(Session::Spectator(_)) => Some(current_frame),
+            None => None,
+        };
+
+        if let Some(confirmed_frame) = confirmed_frame {
+            world.insert_resource(ConfirmedFrameCount(confirmed_frame));
+        }
+
+        if let Some(Session::P2P(s)) = world.get_resource::<Session<T>>() {
+            let confirmed_frame = s.confirmed_frame();
+            world.insert_resource(ConfirmedFrameCount(confirmed_frame));
+        }
```

Comment:
Sorry if you planned to fix this already but the confirmed frame is being set for p2p sessions twice now.

Reply:
Nope I missed that, whoops! Thanks!